### PR TITLE
mypy: make quixstreams.state.* pass type checks

### PR DESCRIPTION
**Pull request by ghost**
mypy: make quixstreams.state.* pass type checks
    
    * Rework message headers type
    * state.partition: remove default from get
    * various type fixes and rework

### CODE REVIEW

File: quixstreams/state/manager.py [hunk 1]
```diff
@@ -230,7 +244,8 @@ def clear_stores(self):
                 "Cannot clear stores with active partitions assigned"
             )
 
-        shutil.rmtree(self._state_dir)
+        if self._state_dir:
```
Comment (gwaramadze) [inline quixstreams/state/manager.py]:
```suggestion
        if self._state_dir is not None:
```

File: quixstreams/models/types.py [hunk 2]
```diff
@@ -1,13 +1,16 @@
-from typing import Dict, List, Optional, Tuple, Union
+from typing import List, Mapping, Optional, Sequence, Tuple, Union
 
 from typing_extensions import Protocol
 
-MessageKey = Union[str, bytes]
+MessageKey = Optional[Union[str, bytes]]
```
Comment (gwaramadze) [inline quixstreams/models/types.py:5]: I think version without `Optional` is better. Using `Optional[MessageKey]` is more explicit and accounts for the possibility that some places will require non-optional `MessageKey`.

File: quixstreams/models/rows.py [hunk 3]
```diff
@@ -19,7 +19,7 @@ def __init__(
         key: Optional[Any],
         timestamp: int,
         context: MessageContext,
-        headers: Optional[MessageHeadersTuples] = None,
+        headers: KafkaHeaders = None,
```
Comment (gwaramadze) [inline quixstreams/models/rows.py:22]: Same here. It's slightly counterintuitive why do I see not `Optional` (or `| None` for later pythons) but default is set to `None`.

File: quixstreams/state/metadata.py [hunk 4]
```diff
@@ -1,13 +1,14 @@
...
 METADATA_CF_NAME = "__metadata__"
 
-Undefined = NewType("Undefined", object)
-UNDEFINED = Undefined(object())
-DELETED = Undefined(object())
-
 DEFAULT_PREFIX = b""
+
+
+class markers(enum.Enum):
```
Comment (gwaramadze) [inline quixstreams/state/metadata.py]: That must not be lowercase plural but uppercase singular. Type hint like `Union[bytes, markers]` must be `Union[bytes, Marker]`.

```suggestion
class Marker(enum.Enum):
```

File: quixstreams/models/types.py (hunk 2 above)
Comment (ghost) [reply quixstreams/models/types.py:5]: I'm a bit torn on this. I keep changing it back and forth. The main argument for including optional in the type is that confluent kafka can have messages keys/values/headers be `None`.

Having a non-optional `MessageKey` is not possible, it's not a message key anymore but a subset.

**Review by gwaramadze — APPROVED** [review]

**Pull request merged**